### DISCUSSION TITLE
Android device cannot be rooted. Alternatives?

### DISCUSSION THREAD

**Discussion by o-alquimista** [wangyu-/udp2raw-tunnel#497] [Q&A] [answered]
My Android device cannot be rooted, thanks to Nokia. Is it possible to run the udp2raw client on my amd64 laptop and somehow use it from my phone?

Or does anyone know an alternative solution for unrooted Android?

**Comment by o-alquimista** [comment, marked as answer]
Solved it.

1. Consider both the laptop and the phone are on the same wifi (campus wifi that blocks wireguard). udp2raw is already running and is usable on my laptop:

```sh
udp2raw -c -l [::]:55820 -r [2001:db8::1]:4096 -k "my_password" --raw-mode faketcp -a
```

> Assume `2001:db8::1` is the address to your udp2raw server instance.

> The listen address is unspecified: `[::]:55820`. This is important, and don't use a weak password!

2. On the phone, in the Wireguard app, edit your connection and set the peer Endpoint to your laptop's local ipv4 or ipv6 address (`ip addr` or `ifconfig`) and the udp2raw port (i'm using 55820).

Ta-da! The university's kung fu is not strong.

### System info
- My laptop is running Fedora Workstation 39, with enforcing SELinux and default firewall configuration. I did not have to change anything.
- The phone is running Android 12.

@wangyu-  It would be nice to add this to the Android guide, right? It's an alternative for people who can't root their phones. It's probably too obvious for most, but I'm a little slow - it took a while to figure this out.

**Reply by wangyu-** [reply]
I have added a link from Andorid Guide pointing to here, since it might not be obvious enough for everyone.

**Reply by ilma17** [reply]
@o-alquimista  did you copy the udp2raw_arm into internal storage. Because From android 12 access of executable from  internal storage gives permission denied issue. I am facing the permission denied issue. It will be a great help if you can give some advice. Thank you.

**Reply by o-alquimista** [reply]
> @o-alquimista did you copy the udp2raw_arm into internal storage. Because From android 12 access of executable from internal storage gives permission denied issue. I am facing the permission denied issue. It will be a great help if you can give some advice. Thank you.

No, I never ran udp2raw on my phone. It ran on my laptop, and my phone just reached it over the network.

I'm sorry I can't be of help. I no longer use udp2raw.

**Reply by Denzy7** [reply]
I have root but actually prefer this method. I feel touching iptables on my stock android phone would mess a lot of android background stuff. Network performance loss is generally negligible and even better I get to offload the computation to the laptop
